Master adp5061 support

stefpopa opened the pull request:
This patch adds basic support for Analog Devices I2C programmable linear
battery charger.

With this driver, some parameters can be read and configured such as:
* trickle charge current level
* trickle charge voltage threshold
* weak charge threshold
* constant current
* constant charge voltage limit
* battery full
* input current limit

Also, commit 413de34 had to be backported because it adds the POWER_SUPPLY_PROP_PRECHARGE_CURRENT property

Next PRs will add extra functionality such as:
* Interrupts support
* Battery pack temperature detection and automatic charger
shutdown
* Fast charge safety timer period
* Watchdog safety timer parameters

commodo commented:
this should be rebased on a newer master ; the git commit range for `checkpatch.pl` is incorrect ;

commodo commented on Documentation/power/power_supply_class.txt:
```diff
@@ -115,28 +115,33 @@ of charge when battery became full/empty". It also could mean "value of
 charge when battery considered full/empty at given conditions (temperature,
 age)". I.e. these attributes represents real thresholds, not design values.
 
+ENERGY_FULL, ENERGY_EMPTY - same as above but for energy.
```
`checkpatch.pl` may raise a warning about the commit comment ; possibly about the commit formatting

stefpopa replied:
this change is part of the backported commit

commodo commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+
+static int adp5061_i2c_write_mask(struct adp5061_state *st,
+				  u8 reg_addr,
+				  unsigned int mask,
```
nitpick/minor: maybe for style consistency, change this to `u32` ; since you are using `u8`

stefpopa replied:
makes sense

mhennerich commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	ADP5061_ILIM_2100mA,
+};
+
+enum adp5061_vterm {
```
If we were to paid by SLOC - I would definitely go for the enum. But I'm not sure if we really need it. And upstream typically prefers denser code. Anyways I'm fine with it one way or the other.

stefpopa replied:
:) ack

mhennerich commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	union {
+		__le16 regval;
+		u8 d8[4];
+	} data ____cacheline_aligned;
```
using regmap we don't need this.

stefpopa replied:
ack

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	st->data.regval |= data;
+
+	ret = regmap_write(st->regmap, reg_addr, st->data.regval);
+
```
Just use regmap_update_bits() instead of re-creating it.

stefpopa replied:
ack; I will use regmap_update_bits()

mhennerich commented on arch/arm/boot/dts/zynq-adrv9361-z7035-box.dts:
```diff
@@ -56,6 +56,11 @@
 		lltc,prescaler-exponent = <5>; /* 2^(2*5) = 1024 */
 	};
+
+	adp5061@14 {
```
Once we merge this into master -

This: https://github.com/analogdevicesinc/rfsom-box-gui/blob/master/bin/batt_man.sh#L34

Will fail - so it needs to be updated to use the ADP5061 driver instead. Please add this to your todo list

stefpopa replied:
I will create a new JIRA ticket for this

rgetz replied:
yeah, this is the only thing - we need to sync this change with the batt_man.sh scripts. For a bit - we may want to see if the device is there (ie, if we are running on a new/old kernel), and either continue to do i2c transactions (if old kernel), or use this device driver (if new kernel).

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	int ret;
+
+	ret = regmap_read(st->regmap, ADP5061_VOLTAGE_TH,
+			  (unsigned int *)&st->data.regval);
```
This won't work on big-endian systems and is not very pretty either way.

For regmap you don't need to use a DMA buffer, regmap takes care of this already. Just put a integer on the stack and use it to read the value.
``` C
unsigned int val;
ret = regmap_read(..., &val);
...
```

Same for all the other places where this pattern is used.

stefpopa replied:
ack

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	struct adp5061_state *st;
+	int ret;
+
+	if (!i2c_check_functionality(adapter, I2C_FUNC_SMBUS_BYTE))
```
I don't think we need this check. We never perform any SMBUS transfers.

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+
+	if (IS_ERR(st->psy)) {
+		dev_err(&client->dev, "Failed to register power supply\n");
+		ret = PTR_ERR(st->psy);
```
return PTR_ERR(...)

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	case POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT:
+	case POWER_SUPPLY_PROP_PRECHARGE_CURRENT:
+	case POWER_SUPPLY_PROP_VOLTAGE_AVG:
+		ret = 1;
```
return 1;

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+		ret = 1;
+		break;
+	default:
+		ret = 0;
```
return 0;

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+		 * above this value, fast chargerge mode is entered
+		 */
+		return adp5061_get_vweak_th(st, val);
+
```
Extra new line

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	if (ret < 0)
+		return ret;
+
+	if (adp5061_chg_type[ADP5061_CHG_STATUS_1_CHG_STATUS(status1)] >
```
Maybe do the `adp5061_chg_type[ADP5061_CHG_STATUS_1_CHG_STATUS(status1)]` once and store it in a helper variable.

stefpopa replied:
makes sense

commodo commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+	return ret;
+}
+
+static int adp5061_get_array_index(struct adp5061_state *st,
```
the state param seems un-needed;
also, i wonder if there is some generic helper for this ; would be nice if there was

commodo commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,851 @@
+		return ret;
+
+	val->intval = adp5061_in_current_lim[
+			      ADP5061_VINX_SET_ILIM_MODE(st->data.regval)];
```
mostly style-wise;
i'd use an extra variable here vs wrapping the line like this;
so:
```
mode = ADP5061_VINX_SET_ILIM_MODE(st->data.regval);
val->intval = adp5061_in_current_lim[mode];
```

stefpopa commented:
changelog v1 -> v2:
* used git commit description style for backported commit
* reduced SLOC by removing unnecessary enums
* used an integer on stack for regmap intead of a DMA buffer
* used regmap_update_bits() instead of i2c_write_msk()
* fixed various style issues
* removed useless if (!i2c_check_functionality(adapter, I2C_FUNC_SMBUS_BYTE)) check
* removed state param from adp5061_get_array_index()
* the power class operates in uV and uA; changed the driver to convert the raw values from micros to millis  
* rebased on the latest master

commodo commented:
you have some warnings from checkpatch.pl
1 requires you to add a doc for the DT binding

lclausen-adi commented:
DT documentation as a separate patch before the driver patch.

stefpopa commented:
changelog v2 -> v3:
* fixed checkpatch.pl warnings by updating MAINTAINERS and adding dt bindings

stefpopa commented:
changelog v3 ->v4:
*fixed typo in commit message

commodo commented on Documentation/devicetree/bindings/power/supply/adp5061.txt:
```diff
@@ -0,0 +1,12 @@
+  - compatible:		should be "adi,adp5061"
+  - reg:		i2c address of the device
+
+Example:
```
this could use a bit more information ;

commodo replied:
maybe check other drivers for reference on what information is preferred to have in docs ;

commodo commented on drivers/power/supply/Makefile:
```diff
@@ -17,6 +17,7 @@ obj-$(CONFIG_WM8350_POWER)	+= wm8350_power.o
 obj-$(CONFIG_TEST_POWER)	+= test_power.o
 
 obj-$(CONFIG_BATTERY_88PM860X)	+= 88pm860x_battery.o
+obj-$(CONFIG_CHARGER_ADP5061)	+= adp5061.o
```
nitpick: this isn't alphabetical ;

commodo replied:
if updating this for alphabetical, make sure to update the order in Kconfig ;

commodo replied:
hmm ; looking through the whole Makefile, there doesn't seem to be a preferred order ;
so, let's leave it as is
sorry for the noise

commodo commented:
rest looks good to me ;

stefpopa commented:
changelog v4->v5:
* added some extra information in the docs

lclausen-adi commented on drivers/power/supply/adp5061.c:
```diff
@@ -0,0 +1,623 @@
+	val /= 1000;
+	index = adp5061_get_array_index(adp5061_in_current_lim,
+					ARRAY_SIZE(adp5061_in_current_lim),
+					val);
```
I'm wondering should we pick the nearest value rather than returning an error when the selected value is not in the table? I'm not sure if there is a preference for either way.

But given that there is no way for an application to enumerate the available values the driver would be really hard to use if the application had to guess by trying different values until it gets no EINVAL.

When picking the nearest value if probably want to make sure that this is the nearest value smaller than the requested limit to make sure the limit is not exceeded.

Same for the other limits.

stefpopa replied:
I will change adp5061_get_array_index() so that it will return the index in the 'array', of greatest value that is less or equal to 'val'

stefpopa commented:
changelog v5 ->v6:
* change adp5061_get_array_index() so that it will return the index in the 'array' of greatest value that is less or equal to 'val'.

stefpopa commented:
changelog v6 ->v7:
* fixed the wrong adp5061_const_ichg[0] 
* fixed adp5061_get_array_index() function

mhennerich commented:
One more comment - the build check only works if we enable the driver in question in our default kernel config.

stefpopa commented:
changelog v7 -> v8:
* enabled adp5061 driver in zynq_xcomm_adv7511_defconfig
